src/Endpoint/Partition.php

```diff
@@ -0,0 +1,163 @@
+<?php
+namespace Aws\Endpoint;
+
+use InvalidArgumentException as Iae;
+
+/**
+ * @internal
+ */
+final class Partition implements \ArrayAccess
+{
+    private $definition;
+
+    public function __construct(array $definition)
+    {
+        foreach (['partition', 'regions', 'services', 'dnsSuffix'] as $key) {
+            if (!isset($definition[$key])) {
+                throw new Iae("Partition missing required $key field");
+            }
+        }
+
+        $this->definition = $definition;
+    }
+
+    public function toArray()
+    {
+        return $this->definition;
+    }
+
+    public function offsetGet($offset)
+    {
+        return isset($this->definition[$offset])
+            ? $this->definition[$offset] : null;
+    }
+
+    public function offsetSet($offset, $value)
+    {
+        $this->definition[$offset] = $value;
+    }
+
+    public function offsetExists($offset)
+    {
+        return isset($this->definition[$offset]);
+    }
+
+    public function offsetUnset($offset)
+    {
+        unset($this->definition[$offset]);
+    }
+
+    /**
+     * Determine if this partition contains the provided region.
+     *
+     * @param string $region
+     *
+     * @return bool
+     */
+    public function matchesRegion($region)
+    {
+        if (isset($this->definition['regionRegex'])) {
```

Comment:
Couldn't you first check if the list of regions for the partition matches the given region, and only then fall back to regex based results?